[SPARK-45205][SQL] CommandResultExec to override iterator methods to avoid triggering multiple jobs.

yorksity opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?

After SPARK-35378 was changed, the execution of statements such as ‘show parititions test' became slower. The change point is that the execution process changes from ExecutedCommandEnec to CommandResultExec, but ExecutedCommandExec originally implemented the following method

override def executeToIterator(): Iterator[InternalRow] = sideEffectResult.iterator

CommandResultExec is not rewritten, so when the hasNext method is executed, a job process is created, resulting in increased time-consuming


### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
Improve performance when show partitions/tables.


### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
No

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->
Existing tests should cover this.

### Was this patch authored or co-authored using generative AI tooling?
<!--
If generative AI tooling has been used in the process of authoring this patch, please include the
phrase: 'Generated-by: ' followed by the name of the tool and its version.
If no, write 'No'.
Please refer to the [ASF Generative Tooling Guidance](https://www.apache.org/legal/generative-tooling.html) for details.
-->
No

yorksity committed:
[SPARK-45205[SQL] CommandResultExec rewritten Iterator method to improve performance when show partitions/tables.

What changes were proposed in this pull request?
After SPARK-35378 was changed, I noticed that the execution of statements such as ‘show parititions test' became slower.
The change point is that the execution process changes from ExecutedCommandEnec to CommandResultExec, but ExecutedCommandExec originally implemented the following method

override def executeToIterator(): Iterator[InternalRow] = sideEffectResult.iterator

CommandResultExec is not rewritten, so when the hasNext method is executed, a job process is created, resulting in increased time-consuming

Why are the changes needed?
Improve performance when show partitions/tables.

Does this PR introduce any user-facing change?
No.

How was this patch tested?
Existing tests should cover this.

Was this patch authored or co-authored using generative AI tooling?
No.

beliefer commented:
@yorksity Good catch! Could you add the running time of ‘show parititions test' before and after this PR ?

beliefer commented:
Please make the title of PR more clear.

yorksity commented:
> Please make the title of PR more clear.

I have modified the title

yorksity commented:
> @yorksity Good catch! Could you add the running time of ‘show parititions test' before and after this PR ?

I created a table using the following statement and added 200 partitions
create table test(id int, name string) partitioned by (groupid int)
When I executed the show partitions test statement in version 3.1.1, it took about 0.2 seconds, while in version 3.3.1, it took about 0.9 seconds and there were jobs generated,
After rewriting the iterator method on CommandResultExec, the statement takes about 0.2 seconds

yorksity commented:
@cloud-fan @yaooqinn @viirya @beliefer  

Please review this PR

beliefer commented:
cc @cloud-fan

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/CommandResultExec.scala:
```diff
@@ -40,6 +40,8 @@ case class CommandResultExec(
 
   override def innerChildren: Seq[QueryPlan[_]] = Seq(commandPhysicalPlan)
 
+  override def executeToIterator(): Iterator[InternalRow] = unsafeRows.iterator
```
good catch! can we follow `ExecutedCommandExec` and put this method right after `def executeCollect`?

yorksity replied:
Modified, thank you for reviewing.

yorksity committed:
[SPARK-45205[SQL] CommandResultExec rewritten Iterator method to improve performance when show partitions/tables.

What changes were proposed in this pull request?
After SPARK-35378 was changed, I noticed that the execution of statements such as ‘show parititions test' became slower.
The change point is that the execution process changes from ExecutedCommandEnec to CommandResultExec, but ExecutedCommandExec originally implemented the following method

override def executeToIterator(): Iterator[InternalRow] = sideEffectResult.iterator

CommandResultExec is not rewritten, so when the hasNext method is executed, a job process is created, resulting in increased time-consuming

Why are the changes needed?
Improve performance when show partitions/tables.

Does this PR introduce any user-facing change?
No.

How was this patch tested?
Existing tests should cover this.

Was this patch authored or co-authored using generative AI tooling?
No.

beliefer reviewed (approved):
LGTM if tests passed.

yorksity committed:
Merge branch 'master' of github.com:apache/spark into SPARK-45205

yorksity commented:
![image](https://github.com/apache/spark/assets/38931534/a28c65ff-fc71-4eff-948f-977efb505c4e)
All tests passed.

cloud-fan commented:
thanks, merging to master/3.5!
